### PR TITLE
Expand auditing observability

### LINKED ISSUE #1459: Expand auditing & observability events
I'd like to expand on the events that can be audited.

Currently [these events are exposed](https://docs.quarkiverse.io/quarkus-langchain4j/dev/observability.html#_auditing):
- initialMessagesCreated
- llmInteractionComplete
- llmInteractionFailed
- responseFromLLMReceived
- toolExecuted

I'd like to add the following auditing events:
- inputGuardrailExecuted
- outputGuardrailExecuted

I'd also like to refactor the event objects. Currently they are Java records, which aren't really good for a public API. Any time we want to add additional information it will be a breaking change. I'd like to introduce an interface for each, then the record can extend the interface. The interface will be part of the public API.

I'd also like to add metrics/tracing/timing to the execution of input & output guardrails.

This is work I can take up. I just wanted to make sure I wasn't duplicating any effort anyone else was working on.

### PR DESCRIPTION
**Pull request by edeandrea**
Expanding auditing & observability

- Closes: #1459

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessor.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+import io.quarkus.arc.deployment.AnnotationsTransformerBuildItem;
+import io.quarkus.deployment.Capabilities;
+import io.quarkus.deployment.Capability;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.deployment.metrics.MetricsCapabilityBuildItem;
+import io.quarkus.runtime.metrics.MetricsFactory;
+
+public class GuardrailExecutionProcessor {
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessor.java:32]: Can we have some Javadoc on what this processor is meant to do?

Also, the name seems a little off, no?

File: core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessor.java [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+                                    AnnotationInstance.builder(MICROMETER_COUNTED)
+                                            .add("value", "guardrail.invoked")
+                                            .add("description",
+                                                    "Measures the number of times this guardrail was invoked")
+                                            .build(),
+                                    //                                    AnnotationInstance.builder(QUARKUS_MICROMETER_COUNTED)
+                                    //                                            .add("value", "guardrail.invoked")
+                                    //                                            .add("description",
+                                    //                                                    "Measures the number of times this guardrail was invoked")
+                                    //                                            .build(),
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessor.java:87]: ?

File: core/deployment/src/test/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+import io.quarkiverse.langchain4j.guardrails.InputGuardrail;
+import io.quarkiverse.langchain4j.guardrails.InputGuardrailParams;
+import io.quarkiverse.langchain4j.guardrails.InputGuardrailResult;
+import io.quarkiverse.langchain4j.guardrails.InputGuardrails;
+import io.quarkiverse.langchain4j.guardrails.OutputGuardrail;
+import io.quarkiverse.langchain4j.guardrails.OutputGuardrailParams;
+import io.quarkiverse.langchain4j.guardrails.OutputGuardrailResult;
+import io.quarkiverse.langchain4j.guardrails.OutputGuardrails;
+
+class GuardrailExecutionProcessorTest {
```
Comment (geoand) [inline core/deployment/src/test/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessorTest.java:39]: We usually don't do these kinds of tests - they really hinder refactoring. We generally use `@QuarkusUnitTest` with a large set of representative "applications"

File: integration-tests/openai/pom.xml [hunk 4]
```diff
@@ -21,6 +21,11 @@
       <artifactId>quarkus-langchain4j-openai</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>io.quarkiverse.langchain4j</groupId>
+      <artifactId>quarkus-langchain4j-core</artifactId>
+      <version>${project.version}</version>
+    </dependency>
```
Comment (geoand) [inline integration-tests/openai/pom.xml:28]: This shouldn't be needed

**Review by geoand — COMMENTED** [review]
Nice, thanks!

I did a first a quick pass, but I'll have to review in depth next week

File: core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessor.java (hunk 1 above)
Comment (edeandrea) [reply core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessor.java:32]: Docs? You want docs? hmmmpf....

Just kidding ;)

File: core/deployment/src/test/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessorTest.java (hunk 3 above)
Comment (edeandrea) [reply core/deployment/src/test/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessorTest.java:39]: I get it (& I did that as well), but since this was pretty complicated class walking I wanted to be able to unit test lots of different edge cases.

Comment (edeandrea) [reply core/deployment/src/test/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessorTest.java]: Just a note - the tests I wrote don't require the `BuildStep`s to run. They're just testing some of the helper methods in the processor class.

Comment (geoand) [reply core/deployment/src/test/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessorTest.java]: > I could refactor those methods out into a helper class if you'd like and then the tests could be associated to the helper class.

That would be nice

**Review by geoand — COMMENTED** [review]
This is really nice work!

I've added a few small things to be addressed, then we can get this one

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/audit/GuardrailExecutedEvent.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Represents an event that is executed when a guardrail validation occurs.
+ * This interface serves as a marker for events that contain both parameters
+ * and results associated with guardrail validation.
+ *
+ * @param <P> the type of guardrail parameters used in the validation process
+ * @param <R> the type of guardrail result produced by the validation process
+ */
+public interface GuardrailExecutedEvent<P extends GuardrailParams, R extends GuardrailResult<R>, G extends Guardrail<P, R>>
```
[the remark this answers is not in the document]
Comment (edeandrea) [reply core/runtime/src/main/java/io/quarkiverse/langchain4j/audit/GuardrailExecutedEvent.java:15]: I tried, but when I did I got all kinds of compile errors:

```
Record 'io.quarkiverse.langchain4j.audit.internal.DefaultInputGuardrailExecutedEvent' from another package not allowed to extend sealed interface 'io.quarkiverse.langchain4j.audit.InputGuardrailExecutedEvent' in unnamed module
```

I'm not sure why, so I decided to leave well enough alone.

[the remark this answers is not in the document]
Comment (geoand) [reply core/runtime/src/main/java/io/quarkiverse/langchain4j/audit/GuardrailExecutedEvent.java:15]: Sure yeah, no complaints on my side :)

File: core/deployment/src/test/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessorTest.java (hunk 3 above)
Comment (edeandrea) [reply core/deployment/src/test/java/io/quarkiverse/langchain4j/deployment/GuardrailExecutionProcessorTest.java]: Done!

**Comment by edeandrea** [comment]
All review comments taken care of.

**Comment by edeandrea** [comment]
@geoand in `GuardrailObservabilityProcessor` I have it split into 2 separate `BuildStep`s - one for the metrics and one for the tracing. I wasn't sure if that was the preferred approach? Or would it be better to just have a single `BuildStep` that did everything?

**Review by geoand — APPROVED** [review]
💪🏽
